File: frontend/src/types/assessment.ts
import type { QualityAnalysis } from './qualityAnalysis'

export type AssessmentType = 'resource' | 'path' | 'topic' | 'test' | 'comprehensive'
export type RecommendationPriority = 'high' | 'medium' | 'low' | string

export interface AssessmentRecommendation {
  title: string
  description?: string
  priority?: RecommendationPriority
  reason?: string
  suggested_action?: string
  related_topics?: string[]
}

export interface AssessmentCreatePayload {
  assessment_type: AssessmentType
  topic?: string | null
  resource_id?: number | null
  path_id?: number | null
  test_id?: number | null
  score?: number | null
  correct_topics?: string[]
  incorrect_topics?: string[]
  learning_evidence?: Record<string, unknown>
}

export interface AssessmentSubmitPayload {
  answers?: Record<string, unknown>
  reflection?: string | null
  self_rating?: number | null
  feedback?: string | null
}

export interface AssessmentItem {
  id: number
  assessment_id?: number | null
  title?: string | null
  assessment_type: AssessmentType
  target_type?: string | null
  target_id?: number | null
  topic?: string | null
  resource_id?: number | null
  path_id?: number | null
  test_id?: number | null
  score?: number | null
  level?: string | null
  summary?: string | null
  strengths?: string[]
  weaknesses?: string[]
  weak_topics?: string[]
  correct_topics: string[]
  incorrect_topics: string[]
  analysis?: string | null
  recommendations: AssessmentRecommendation[]
  answers?: Record<string, unknown>
  reflection?: string | null
  self_rating?: number | null
  feedback?: string | null
  quality_analysis?: QualityAnalysis | null
  created_at: string
  updated_at?: string | null
  submitted_at?: string | null
}

export interface AssessmentListResponse {
  items: AssessmentItem[]
  total: number
  page: number
  page_size: number
}

export interface AssessmentTrendPoint {
  assessment_id: number
  score?: number | null
  created_at: string
}

export interface AssessmentSummary {
  total_assessments: number
  average_score: number
  latest_score?: number | null
  score_trend: AssessmentTrendPoint[]
  strong_topics: string[]
  weak_topics: string[]
  assessment_type_distribution: Record<string, number>
  recent_recommendations: AssessmentRecommendation[]
  profile_dimension_hints: Record<string, number>
}

export interface AssessmentRecommendationResponse {
  recommendations: AssessmentRecommendation[]
  basis: Record<string, unknown>
}

export interface AssessmentQuery {
  page?: number
  page_size?: number
  assessment_type?: AssessmentType | ''
  topic?: string
  min_score?: number | null
  max_score?: number | null
}

File: backend/app/schemas/assessment.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.quality_analysis import QualityAnalysis


AssessmentType = Literal["resource", "path", "topic", "test", "comprehensive"]
RecommendationPriority = Literal["high", "medium", "low"]


class AssessmentRecommendation(BaseModel):
    title: str
    description: str
    priority: RecommendationPriority = "medium"
    reason: str | None = None
    suggested_action: str
    related_topics: list[str] = Field(default_factory=list)


class LearningAssessmentCreate(BaseModel):
    assessment_type: AssessmentType = Field(description="Assessment scope")
    topic: str | None = Field(default=None, max_length=160)
    resource_id: int | None = Field(default=None, gt=0)
    path_id: int | None = Field(default=None, gt=0)
    test_id: int | None = Field(default=None, gt=0)
    score: float | None = Field(default=None, ge=0, le=100)
    correct_topics: list[str] = Field(default_factory=list)
    incorrect_topics: list[str] = Field(default_factory=list)
    learning_evidence: dict[str, Any] = Field(default_factory=dict)


class LearningAssessmentSubmit(BaseModel):
    answers: dict[str, Any] = Field(default_factory=dict)
    reflection: str | None = Field(default=None, max_length=4000)
    self_rating: float | None = Field(default=None, ge=0, le=100)
    feedback: str | None = Field(default=None, max_length=4000)


class LearningAssessmentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    assessment_id: int | None = None
    title: str | None = None
    assessment_type: str
    target_type: str | None = None
    target_id: int | None = None
    topic: str | None = None
    resource_id: int | None = None
    path_id: int | None = None
    test_id: int | None = None
    score: float | None = None
    level: str | None = None
    summary: str | None = None
    strengths: list[Any] = Field(default_factory=list)
    weaknesses: list[Any] = Field(default_factory=list)
    weak_topics: list[Any] = Field(default_factory=list)
    correct_topics: list[Any] = Field(default_factory=list)
    incorrect_topics: list[Any] = Field(default_factory=list)
    analysis: str | None = None
    recommendations: list[AssessmentRecommendation] = Field(default_factory=list)
    answers: dict[str, Any] = Field(default_factory=dict)
    reflection: str | None = None
    self_rating: float | None = None
    feedback: str | None = None
    quality_analysis: QualityAnalysis | None = None
    created_at: datetime
    updated_at: datetime
    submitted_at: datetime | None = None


class LearningAssessmentListResponse(BaseModel):
    items: list[LearningAssessmentRead]
    total: int
    page: int
    page_size: int


class AssessmentTrendPoint(BaseModel):
    assessment_id: int
    score: float | None = None
    created_at: datetime


class LearningAssessmentSummary(BaseModel):
    total_assessments: int
    average_score: float
    latest_score: float | None = None
    score_trend: list[AssessmentTrendPoint] = Field(default_factory=list)
    strong_topics: list[str] = Field(default_factory=list)
    weak_topics: list[str] = Field(default_factory=list)
    assessment_type_distribution: dict[str, int] = Field(default_factory=dict)
    recent_recommendations: list[AssessmentRecommendation] = Field(default_factory=list)
    profile_dimension_hints: dict[str, float] = Field(default_factory=dict)


class LearningRecommendationResponse(BaseModel):
    recommendations: list[AssessmentRecommendation]
    basis: dict[str, Any]
